utils/utils/src/main/scala/pl/touk/nussknacker/engine/util/cache/LazyMap.scala

```diff
@@ -0,0 +1,75 @@
+package pl.touk.nussknacker.engine.util.cache
+
+import cats.implicits._
+import com.github.benmanes.caffeine.cache.{CacheLoader, Caffeine, LoadingCache}
+
+import java.util.concurrent.CompletionException
+import java.util.stream.Collectors
+import java.util.stream.Collectors.toList
+import java.{util => ju}
+import scala.jdk.CollectionConverters._
+import scala.util.Try
+
+class LazyMap[A <: AnyRef, B >: Null](
+    definedFields: ju.Set[A],
+    extractValue: A => B
+) extends ju.AbstractMap[A, B]
+    with ju.Map[A, B] {
+
+  import LazyMap._
+
+  private val loader: CacheLoader[A, Option[B]] = key => Some(extractValue(key))
+  private val cache: LoadingCache[A, Option[B]] = Caffeine.newBuilder.build(loader)
```

Comment:
I'm wondering isn't this a problem for Flink serialization? I mean, don't we need here `@Transient`?

Reply:
We don't need `LoadingCache`, it's too heavy and the only thing it gives us is nicer API at the cost of more memory use This builder creates `UnboundedLocalCache.UnboundedLocalLoadingCache` - basically a wrapper around `ConcurrentHashMap`. So let's use `ConcurrentHashMap` directly.